src/blockstore/hashmap_db.zig

```diff
@@ -0,0 +1,237 @@
+const std = @import("std");
+const sig = @import("../lib.zig");
+
+const Allocator = std.mem.Allocator;
+const DefaultRwLock = std.Thread.RwLock.DefaultRwLock;
+
+const BytesRef = sig.blockstore.database.BytesRef;
+const ColumnFamily = sig.blockstore.database.ColumnFamily;
+const Logger = sig.trace.Logger;
+
+const serializeAlloc = sig.blockstore.database.serializer.serializeAlloc;
+const deserialize = sig.blockstore.database.serializer.deserialize;
+
+pub fn SharedHashMapDB(comptime column_families: []const ColumnFamily) type {
+    return struct {
+        allocator: Allocator,
+        maps: []SharedHashMap,
+        /// shared lock is required to call locking map methods.
+        /// exclusive lock is required to call non-locking map methods.
+        transaction_lock: DefaultRwLock = .{},
+
+        const Self = @This();
+
+        pub fn open(
+            allocator: Allocator,
+            _: Logger,
+            _: []const u8,
+        ) Allocator.Error!Self {
+            var maps = try allocator.alloc(SharedHashMap, column_families.len);
+            errdefer {
+                for (maps) |*m| m.deinit();
+                allocator.free(maps);
+            }
+            inline for (0..column_families.len) |i| {
+                maps[i] = try SharedHashMap.init(allocator);
+            }
+            return .{ .allocator = allocator, .maps = maps };
+        }
+
+        pub fn deinit(self: *Self) void {
+            for (self.maps) |*map| {
+                map.deinit();
+            }
+            self.allocator.free(self.maps);
+        }
+
+        pub fn put(
+            self: *Self,
+            comptime cf: ColumnFamily,
+            key: cf.Key,
+            value: cf.Value,
+        ) Allocator.Error!void {
+            const key_bytes = try serializeAlloc(self.allocator, key);
+            errdefer self.allocator.free(key_bytes);
```

Comment:
can we add some spaces between after the `errdefer`s ? hard to read whats going on

Reply:
8a89d582b72043deaec934200e7150fd61c14086